packages/babel-helper-builder-react-jsx-experimental/src/index.js

```diff
@@ -138,53 +151,48 @@ You can set \`throwIfNamespace: false\` to bypass this warning.`,
 
         // For jsx mode
         let source = IMPORT_SOURCE_DEFAULT;
-        let sourceSet = !!options.importSource;
 
         // For createElement mode
         let pragma = PRAGMA_DEFAULT;
         let pragmaFrag = PRAGMA_FRAG_DEFAULT;
-        let pragmaSet = !!options.pragma;
-        let pragmaFragSet = !!options.pragmaFrag;
+
+        let runtimeMatch;
+        let sourceMatch;
+        let jsxMatch;
+        let jsxFragMatch;
 
         if (file.ast.comments) {
           for (const comment of (file.ast.comments: Array<Object>)) {
-            const sourceMatches = JSX_SOURCE_ANNOTATION_REGEX.exec(
-              comment.value,
-            );
-            if (sourceMatches) {
-              source = sourceMatches[1];
-              sourceSet = true;
-            }
-
-            const runtimeMatches = JSX_RUNTIME_ANNOTATION_REGEX.exec(
-              comment.value,
-            );
-            if (runtimeMatches) {
-              runtime = runtimeMatches[1];
-            }
-
-            const jsxMatches = JSX_ANNOTATION_REGEX.exec(comment.value);
-            if (jsxMatches) {
-              pragma = jsxMatches[1];
-              pragmaSet = true;
-            }
-            const jsxFragMatches = JSX_FRAG_ANNOTATION_REGEX.exec(
-              comment.value,
-            );
-            if (jsxFragMatches) {
-              pragmaFrag = jsxFragMatches[1];
-              pragmaFragSet = true;
-            }
+            runtimeMatch =
+              JSX_RUNTIME_ANNOTATION_REGEX.exec(comment.value)?.[1] ??
+              runtimeMatch;
+
+            sourceMatch =
+              JSX_SOURCE_ANNOTATION_REGEX.exec(comment.value)?.[1] ??
+              sourceMatch;
+
+            jsxMatch =
+              JSX_ANNOTATION_REGEX.exec(comment.value)?.[1] ?? jsxMatch;
+
+            jsxFragMatch =
+              JSX_FRAG_ANNOTATION_REGEX.exec(comment.value)?.[1] ??
+              jsxFragMatch;
           }
         }
 
+        if (runtimeMatch) {
+          runtime = runtimeMatch;
+        } else if (sourceMatch) {
+          runtime = "automatic";
+        } else if (jsxMatch || jsxFragMatch) {
+          runtime = "classic";
+        }
```

Comment:
I think that if `runtimeMatch === "classic" && (jsxMatch || jsxFragMatch)` or if `runtimeMatch === "automatic" && sourceMatch` we should throw an error, since you are setting two incompatible options at the same level.

Reply:
Also, what happens if I set both `@jsxImportSource` and `@jsxPragma` in the same file?

Reply:
> I think that if runtimeMatch === "classic" && (jsxMatch || jsxFragMatch) or if runtimeMatch === "automatic" && sourceMatch we should throw an error, since you are setting two incompatible options at the same level.

Sure, can add that - a question is: do you plan to keep `@jsxRuntime` in Babel 8? I somewhat believe this is redundant and not needed.